src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -185,121 +166,188 @@ def __enter__(self):
-                continue
+                self._batch_start_time = datetime.utcnow()
+                # After starting time, put the input in self._task_queue, because the monitor thread will
+                # use self._batch_start_time to calculate the remain execution time after get the input data.
+                for index, inputs in batch_inputs:
+                    self._task_queue.put(
+                        (
+                            self._run_id,
+                            index,
+                            inputs,
+                        )
+                    )
+                # Wait for batch run to complete or timeout
+                completed_line = 0
+                while not self._batch_timeout_expired(self._batch_start_time):
+                    # Print the progress logs of the batch run.
+                    log_progress(
+                        run_start_time=self._batch_start_time,
+                        logger=bulk_logger,
+                        count=len(self._result_dict),
+                        total_count=self._nlines,
+                        last_log_count=completed_line,
+                    )
+                    completed_line = len(self._result_dict)
+                    # If the batch run is completed, break the loop.
+                    if self._is_batch_run_completed():
+                        break
+                    # Check monitor status every 1 second
+                    self._monitor_thread_pool_status()
+                    await asyncio.sleep(1)
+            except PromptflowException:
+                raise
             except Exception as e:
-                raise Exception(f"Unexpected error occurred while get process info. Exception: {e}")
-
-    def _ensure_process_terminated_within_timeout(self, process_id):
-        start_time = time.time()
-        while psutil.pid_exists(process_id):
-            if time.time() - start_time > self._PROCESS_TERMINATED_TIMEOUT:
-                raise ProcessTerminatedTimeout(self._PROCESS_TERMINATED_TIMEOUT)
-            time.sleep(1)
+                bulk_logger.error(f"ProcessPool failed with exception: {e}")
+                raise ProcessPoolError(
+                    message_format=f"ProcessPool failed with exception: {e}",
+                    target=ErrorTarget.EXECUTOR,
+                ) from e
 
-    def _is_process_alive(self, process_id):
-        return psutil.pid_exists(process_id)
+        if self._batch_timeout_expired(self._batch_start_time):
+            # Send terminate signal to all threads and wait for them to exit.
+            self._terminate_tasks()
+            # Wait for up to 10s for thread termination, aiming to ensure that
+            # the line results in the result dict are as complete as possible.
+            start_time = datetime.utcnow()
+            while not self._timeout_expired(start_time, self._THREAD_TERMINATED_TIMEOUT):
+                if self._all_tasks_ready():
+                    break
```

Comment:
Better to add time.sleep(1) to avoid memory brust.

Reply:
Updated.